src/middlewared/middlewared/pytest/unit/utils/test_passdb.py

```diff
@@ -0,0 +1,217 @@
+    """
+    expected = set(usernames)
+    found = []
+    pdbedit = subprocess.run(['pdbedit', '-L'], capture_output=True)
+    assert pdbedit.returncode == 0, pdbedit.stderr.decode()
+    for entry in pdbedit.stdout.decode().strip().splitlines():
+        found.append(entry.split(':')[0])
+
+    assert len(found) == len(usernames), f'expected: {usernames}, found: {found}'
+
+    found = set(found)
+    missing = expected - found
+    assert missing == set(), str(missing)
+
+    extra = found - expected
+    assert extra == set(), str(extra)
+
+
+@pytest.mark.parametrize('pdbentrydict', [
+    PDB_DICT_DEFAULTS | {'username': 'user1', 'full_name': 'user1', 'user_rid': 20071},
+    PDB_DICT_DEFAULTS | {'username': 'user2', 'full_name': 'user2', 'user_rid': 20072, 'acct_ctrl': LOCKED_ACCOUNT},
+    PDB_DICT_DEFAULTS | {'username': 'user3', 'full_name': 'user3', 'user_rid': 20073, 'acct_ctrl': DISABLED_ACCOUNT},
+    PDB_DICT_DEFAULTS | {'username': 'user4', 'full_name': 'user4', 'user_rid': 20074, 'acct_ctrl': EXPIRED_ACCOUNT},
+])
+def test__insert_query_delete(pdbentrydict, passdb_dir, pdb_times):
+    """ Add user, verify it shows in our query, verify that it shows in pdbedit, then delete it """
+
+    # generate NT hash from randomized password
+    nt_hash = generate_nt_hash(''.join(secrets.choice(string.ascii_letters + string.digits) for i in range(10)))
+    payload = pdbentrydict | {'times': pdb_times, 'nt_pw': nt_hash}
+
+    entry = util_passdb.PDBEntry(**payload)
+
+    util_passdb.insert_passdb_entries([entry])
+
+    try:
+        contents = util_passdb.query_passdb_entries([], {})
+        assert len(contents) == 1
+        check_pdbedit([entry.username])
+    finally:
+        util_passdb.delete_passdb_entry(payload['username'], payload['user_rid'])
+
+    assert asdict(entry) == contents[0]
+
+    contents = util_passdb.query_passdb_entries([], {})
+    assert len(contents) == 0, str(contents)
+
+
+def test__smbhash_parser():
+    """
+    The `smbhash` field in samba has changed from being a smbpasswd string to simply
+    containing the NT hash for user. This test ensures that we properly extract NT hash
+    from both entry types.
+    """
+    smb_hash_legacy = "smbuser:3000:XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX:05BC65787F63B56CF6D47F16E32E3ABF:[U         ]:LCT-66C7190E:"  # noqa
+    nt_hash = "05BC65787F63B56CF6D47F16E32E3ABF"
+
+    assert util_passdb.user_smbhash_to_nt_pw('smbuser', smb_hash_legacy) == nt_hash
+    assert util_passdb.user_smbhash_to_nt_pw('smbuser', nt_hash) == nt_hash
+
+
+@pytest.mark.parametrize('user_data,expected', [
+    ({'locked': False, 'password_disabled': False}, util_passdb.UserAccountControl.NORMAL_ACCOUNT),
+    ({'locked': True, 'password_disabled': False}, LOCKED_ACCOUNT),
+    ({'locked': False, 'password_disabled': True}, DISABLED_ACCOUNT),
+    ({'locked': True, 'password_disabled': True}, DISABLED_ACCOUNT | LOCKED_ACCOUNT),
+])
```

Comment:
Nothing wrong with this, but wondering why `user_entry_to_uac_flags` doesn't also handle `DONT_EXPIRE_PASSWORD`.  If it's not applicable maybe add a comment to that effect.

Reply:
We don't have a corresponding user field _yet_.